Add transformer model to diffusion policy

amandip7 opened the pull request:
## What this does
Implements transformer model for diffusion policy : https://github.com/orgs/huggingface/projects/46?pane=issue&itemId=64857791
This adds a transformer model class to modelling_diffusion, adds configurations required for this model and sets the default model in diffusion to be transformer instead of U-net.

## How it was tested
- Trained a transformer diffusion policy with the default configurations provided.
- Ported a pre-trained transformer diffusion policy from original DP repo to lerobot and ran evaluation here.

Examples:
- Added `test_something` in `tests/test_stuff.py`.
- Added `new_feature` and checked that training converges with policy X on dataset/environment Y.
- Optimized `some_function`, it now runs X times faster than previously.

## How to checkout & try? (for the reviewer)
Run the training script:
python lerobot/scripts/train.py
This would train the transformer based diffusion policy with default configuration over pushT task.

alexander-soare commented on lerobot/common/policies/diffusion/configuration_diffusion.py:
```diff
@@ -98,7 +98,7 @@ class DiffusionConfig:
     # Inputs / output structure.
     n_obs_steps: int = 2
-    horizon: int = 16
+    horizon: int = 10
```
I think this change should be reverted right? To keep in line with the default PushT policy.

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/configuration_diffusion.py:
```diff
@@ -134,7 +134,7 @@ class DiffusionConfig:
     kernel_size: int = 5
     n_groups: int = 8
-    diffusion_step_embed_dim: int = 128
+    diffusion_step_embed_dim: int = 256
```
I think this change should be reverted right? To keep in line with the default PushT policy.

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        self.config = config
+
+        # compute number of tokens for main trunk and condition encoder
+        if config.n_obs_steps is None:
```
config.n_obs_steps is None does not seem to be allowed according to the type hinting and documentation. So perhaps it doesn't make sense to handle it here, right?

amandip7 replied:
Right. This has been removed now

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        if config.n_obs_steps is None:
+            config.n_obs_steps = config.horizon
+
+        t = config.horizon
```
Hmm, is it okay if we just leave this as config.horizon rather than binding it to another much less descriptive variable name?

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+            config.n_obs_steps = config.horizon
+
+        t = config.horizon
+        t_cond = 1
```
So what is `t_cond`? Could you please use a more descriptive variable name or if that's not appropriate, just leave a comment here?

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        self.drop = nn.Dropout(config.p_drop_emb)
+
+        # cond encoder
+        self.time_emb = DiffusionSinusoidalPosEmb(config.diffusion_step_embed_dim)
```
`diffusion_step_embed_dim` is found under the `# Unet` part of the config. Could you please consider making a `# Architecture shared params` section as well as `# Unet` and `# Transformer`? (or whatever else you think is suitable to make the configuration params clearer.

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+
+        # cond encoder
+        self.time_emb = DiffusionSinusoidalPosEmb(config.diffusion_step_embed_dim)
+        self.cond_obs_emb = None
```
This line appears to be redundant.

amandip7 replied:
Cleared this code

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+
+        self.cond_obs_emb = nn.Linear(cond_dim, config.diffusion_step_embed_dim)
+
+        self.cond_pos_emb = None
```
This line appears to be redundant.

amandip7 replied:
Cleared

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+
+
+class TransformerForDiffusion(nn.Module):
+    def __init__(self, config: DiffusionConfig, cond_dim: int):
```
May I please ask for either more descriptive variable names or comments describing what the variables mean here? I've highlighted at least 1 or 2 specific asks below, but I realized it might be better to make this general request. Please check ACT for inspiration

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        self.cond_obs_emb = nn.Linear(cond_dim, config.diffusion_step_embed_dim)
+
+        self.cond_pos_emb = None
+        self.encoder = None
```
These two lines also.

amandip7 replied:
Cleared these

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+
+        # init
+        self.apply(self._init_weights)
+        # logger.info(
```
Could you please remove this commented code?

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        #     "number of parameters: %e", sum(p.numel() for p in self.parameters())
+        # )
+
+    def _init_weights(self, module):
```
Is there anything we can say in a docstring to summarize the weight initialization strategy used here?

amandip7 replied:
Docstring is added for this method now

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        cond_embeddings = torch.cat([cond_embeddings, cond_obs_emb], dim=1)
+        # (B,To + 1,n_emb)
+
+        tc = cond_embeddings.shape[1]
```
`tc` is used only one and on the next line. In this instance I think it would make more sense to just use the RHS of this assignment in the line below, rather than adding an obscure variable name into the namespace.

amandip7 replied:
Removed tc variable.

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        return optimizer
+
+    def forward(self, sample: torch.Tensor, timestep: torch.Tensor, global_cond: torch.Tensor, **kwargs):
+        """
```
Could we please tidy up this docstring?
- make sure documented variable names match the function signature
- Explain what the the variables are where not obvious, expecially timestamp (see `DiffusionConditionalUnet1d.forward`.

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        output: (B,T,input_dim)
+        """
+        # 1. time
+        timesteps = timestep
```
Any reason you assign another variable to the same object, and with such a similar name?

amandip7 replied:
Cleared this. We donot need to assign to a new variable here.

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        # 1. time
+        timesteps = timestep
+        batch_size = sample.shape[0]
+        # broadcast to batch dimension in a way that's compatible with ONNX/Core ML
```
We don't need this at the moment.

amandip7 replied:
Cleared.

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        input_emb = self.input_emb(sample)
+
+        # encoder
+        cond_embeddings = time_emb
```
nit: mind dropping this line and just putting `time_emb` into `torch.cat` instead?

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        x = self.drop(cond_embeddings + position_embeddings)
+        x = self.encoder(x)
+        memory = x
+        # (B,T_cond,n_emb)
```
Can we make comments like this either in-line, or on the line _preceding_ the line of code of concern? I think putting code then comment on the next line is rather unconventional.

amandip7 replied:
Done

alexander-soare commented on lerobot/common/policies/diffusion/modeling_diffusion.py:
```diff
@@ -749,3 +768,264 @@ def forward(self, x: Tensor, cond: Tensor) -> Tensor:
+        position_embeddings = self.cond_pos_emb[:, :tc, :]  # each position maps to a (learnable) vector
+        x = self.drop(cond_embeddings + position_embeddings)
+        x = self.encoder(x)
+        memory = x
```
Do we really have to add another variable into the namespace here?

amandip7 replied:
Cleared this variable

alexander-soare reviewed (changes requested):
Thanks @amandip7 for this PR! People have been asking about the transformer version of diffusion policy for a while.

I couple of high-level comments to go with my first review:
- What do you say that we add a `diffusion_transformer.yaml` instead of updating the current `diffusion.yaml` to use the transformer variant?
- Can you push the weights to the hub so I can test it out with an eval run? Can you please update the PR notes to include instructions on how to run the eval, and what results I should expect? And can you confirm that the success rate is the same as that reported on the paper and/or the same as you get when you run eval on the original DP repo?
- Have you confirmed that you can train a model on LeRobot and achieve the same success rate as the ported weights? If so, can you pleas push a model to the hub and add instructions on how to evaluate that one?
- Most of my in-line comments are around tidying up the code a little. I understand you've probably copied the code directly from the DP repo, but we'd like to make the code on LeRobot a little more accessible. This usually means tidying it up a bit by removing redundant code, adding comments, and using more self-explanatory variable names.

Thanks again!

mohitydv09 commented:
Hey @amandip7 , are you working on this still or can I take up this task??

amandip7 commented:
> Hey @amandip7 , are you working on this still or can I take up this task??

Hey @mohitydv09 Im working on it.

mohitydv09 commented:
@amandip7  Ohh okay, let me know if you want to delegate some sub task in the PR. I will be implementing the Transformers based Diffusion Policy for my research work anyways so I can take up some work if you want. :-)

amandip7 commented:
Thanks for the review @alexander-soare ! I have cleared the code and addressed your comments. A new config file diffusion_transformer.yaml is now used for transformer model. diffusion.yaml remains unchanged.
### To Checkout and try
The default config for the transformer model trains a model with same configuration as in the DP paper. 
A pretrained model from DP repo is converted to lerobot and evaluated here. We have also trained a transformer based diffusion policy in lerobot with the same configuration. The model weights and evaluation results are pushed to the hub along with weights conversion script : https://huggingface.co/Amandee/diffusion_transformer_pusht
We achieve results comparable to the pretrained original DP repo's model through a model trained within lerobot using this implementation.

vaishanth-rmrj commented:
@amandip7 Thanks for implementing the transformer backbone. Have you tried training and evaluating on any of the robot arms? Any luck getting it to work ?

amandip7 commented:
> @amandip7 Thanks for implementing the transformer backbone. Have you tried training and evaluating on any of the robot arms? Any luck getting it to work ?

I trained it on the PushT task and evaluated here. It's not trained with an actual robot arm yet @vaishanth-rmrj

vaishanth-rmrj commented:
@amandip7 I tried training it on my So-100 arm dataset. The policy training flow works without throwing any errors. But the policy is not successfully able to perform the task, I guess this has nothing to do with the implementation or the policy. I am still experimenting with different policy configurations. I have noticed the same with the Diffusion policy with Unet backbone. 

Wondering why it still isn't pushed to the main repo ?

amandip7 commented:
> @amandip7 I tried training it on my So-100 arm dataset. The policy training flow works without throwing any errors. But the policy is not successfully able to perform the task, I guess this has nothing to do with the implementation or the policy. I am still experimenting with different policy configurations. I have noticed the same with the Diffusion policy with Unet backbone.
> 
> Wondering why it still isn't pushed to the main repo ?

Hey @vaishanth-rmrj one thing to notice is that the default policy that this trains is about 20M params while the default policy the Unet mode trains is 10x the size. Hence this performs a little worse. The reason for this size is to be in sync with original DP repo. You can find more details comparing the models here :
Transformer : https://huggingface.co/Amandee/diffusion_transformer_pusht
Unet : https://huggingface.co/lerobot/diffusion_pusht

Why it's not Pushed: I'm looking for code review and feedback. Could use your help here @aliberts

vaishanth-rmrj commented:
@amandip7 I conducted experiments comparing transformer-based and U-Net-based DP models. The results indicate that the transformer-backbone model outperforms the U-Net model, even when using a smaller architecture.

@aliberts Could you please review this PR and provide your feedback? 

# Experiment Details

## Overview
- **Episodes:** 200  
- **Iterations:** 80,000  
- **Training Timesteps per Episode:** 10  
- **Inference Steps per Episode:** 10  
- **Horizon:** 100  
- **Action Steps:** 100  
- **Observation Steps:** 4  

## Input Shapes
- **Observation Images:** `[3, 480, 640]` (Important)

## Models and Success Rates
| Model Description                          | Parameters | Success Rate |
|-------------------------------------------|------------|--------------|
| DP with Transformer Backbone              | 26M        | 72%          |
| DP with U-Net                             | 265M       | 58%          |
| DP with U-Net                             | 17M        | 56%          |


# Demo Video for Transformer DP

https://github.com/user-attachments/assets/c14abd31-ed05-4354-94c5-a1b0f02f23e7

# Config Used for Transformer DP
[diffusion_transformer_so100.txt](https://github.com/user-attachments/files/19079225/diffusion_transformer_so100.txt)

liuhuanjim013 commented:
@vaishanth-rmrj I've merged latest main into this branch and created a [PR](https://github.com/amandip7/lerobot/pull/1). I used it to train and run evaluation. Could you try to see if it also works for you?

kappa95 commented:
I am trying to use lerobot diffusion policy with another robot and compared to act i do not have so much success. Do you have a sample infer script for diffusion policy? This would be very helpful
